tests/components/template/test_sensor.py

```diff
@@ -438,6 +443,45 @@ def test_available_template_with_entities(self):
         )
 
 
+async def test_creating_sensor_loads_group(hass):
+    """Test setting up template sensor loads group component first."""
+    order = []
+    after_dep_event = Event()
+
+    async def async_setup_group(hass, config):
+        # Make sure group takes longer to load, so that it won't
+        # be loaded first by chance
+        await after_dep_event.wait()
+
+        order.append("group")
+        return True
+
+    async def async_setup_template(
+        hass, config, async_add_entities, discovery_info=None
+    ):
+        order.append("sensor.template")
+        return True
+
+    async def set_after_dep_event(event):
+        if event.data[ATTR_COMPONENT] == "sensor":
+            after_dep_event.set()
```

Comment:
Can we skip patching `async_setup` and `async_setup_platform` and push the name of the `ATTR_COMPONENT` into `order`, call `hass.async_block_till_done`, and then check `order`?

Reply:
I tried, but it didn't work.

This test is green with or without `after_dependencies` in manifest.json:
```python
async def test_creating_sensor_loads_group(hass):
    """Test setting up template sensor loads group component first."""
    order = []

    async def async_add_to_order(event: Event) -> None:
        order.append(event.data[ATTR_COMPONENT])

    hass.bus.async_listen(EVENT_COMPONENT_LOADED, async_add_to_order)

    await async_from_config_dict(
        {"sensor": {"platform": "template", "sensors": {}}, "group": {}}, hass
    )
    await hass.async_block_till_done()

    assert ["group", "sensor"] == order[2:]
```

This test is red with or without `after_dependencies` in manifest.json, as `order` comes back as `["sensor", "group"]`:

```python
async def test_creating_sensor_loads_group(hass):
    """Test setting up template sensor loads group component first."""
    order = []

    async def async_add_to_order(event: Event) -> None:
        order.append(event.data[ATTR_COMPONENT])

    hass.bus.async_listen(EVENT_COMPONENT_LOADED, async_add_to_order)

    async def delay_setup(hass, config):
        # Make sure group takes longer to load, so that it won't
        # be loaded first by chance
        await sleep(0.1)

        return True

    with patch(
        "homeassistant.components.group.async_setup", new=delay_setup,
    ):
        await async_from_config_dict(
            {"sensor": {"platform": "template", "sensors": {}}, "group": {}}, hass
        )
        await hass.async_block_till_done()

    assert ["group", "sensor"] == order[2:]
```

The log messages look like `sensor.template` is loaded after Event component_loaded is emitted:
```
INFO     homeassistant.loader:loader.py:214 Loaded sensor from homeassistant.components.sensor
INFO     homeassistant.loader:loader.py:214 Loaded group from homeassistant.components.group
INFO     homeassistant.bootstrap:bootstrap.py:457 Domains to be set up: {'group', 'sensor'}
INFO     homeassistant.bootstrap:bootstrap.py:514 Setting up stage 2: {'group', 'sensor'}
INFO     homeassistant.setup:setup.py:162 Setting up group
INFO     homeassistant.loader:loader.py:214 Loaded template from homeassistant.components.template
INFO     homeassistant.setup:setup.py:162 Setting up sensor
DEBUG    homeassistant.setup:setup.py:91 Dependency template will wait for ['group']
INFO     homeassistant.setup:setup.py:212 Setup of domain sensor took 0.0 seconds
DEBUG    homeassistant.core:core.py:587 Bus:Handling <Event component_loaded[L]: component=sensor>
INFO     homeassistant.setup:setup.py:212 Setup of domain group took 0.1 seconds
DEBUG    homeassistant.core:core.py:587 Bus:Handling <Event component_loaded[L]: component=group>
DEBUG    homeassistant.bootstrap:bootstrap.py:518 Waiting for startup to wrap up
INFO     homeassistant.components.sensor:entity_platform.py:166 Setting up sensor.template
INFO     homeassistant.bootstrap:bootstrap.py:233 Home Assistant initialized in 0.12s
```

Reply:
Maybe we should emit an `<Event platform_loaded platform=sensor.template>` in `EntityPlatform._async_setup_platform`?

Reply:
Ideally we don't add any more events as they can slow things down.

Can we look at the order of  `hass.config.components`?

Reply:
`hass.config.components` seems to be a set, so the order is random on each test run. I've had runs where `group` was the first element of `list(hass.config.components)`, and other runs where it was the last element.